### ISSUE TITLE
SI-1980 by-name argument incorrectly evaluated on :-ending operator

### ISSUE THREAD

**Issue by xuwei-k** [scalaz/scalaz#528]
https://issues.scala-lang.org/browse/SI-1980

for example

```
scala> import scalaz._,Scalaz._
import scalaz._
import Scalaz._

scala> {println("failure")} <<?: 1.failure[String] :?>> {println("success")}
failure

scala> {println("failure")} <<?: 1.success[String] :?>> {println("success")}
failure
success
```
- [`FingerTree#+:`](https://github.com/scalaz/scalaz/blob/v7.1.0-M2/core/src/main/scala/scalaz/FingerTree.scala#L390)
- [`FingerTree#|-:`](https://github.com/scalaz/scalaz/blob/v7.1.0-M2/core/src/main/scala/scalaz/FingerTree.scala#L412)
- [`IndSeq#+:`](https://github.com/scalaz/scalaz/blob/v7.1.0-M2/core/src/main/scala/scalaz/FingerTree.scala#L1002)
- [`EitherT#<<?:`](https://github.com/scalaz/scalaz/blob/v7.1.0-M2/core/src/main/scala/scalaz/EitherT.scala#L18)
- [`Either#<<?:`](https://github.com/scalaz/scalaz/blob/v7.1.0-M2/core/src/main/scala/scalaz/Either.scala#L8)
- [`StreamT#::`](https://github.com/scalaz/scalaz/blob/v7.1.0-M2/core/src/main/scala/scalaz/StreamT.scala#L19)
- [`Validation#<<?:`](https://github.com/scalaz/scalaz/blob/v7.1.0-M2/core/src/main/scala/scalaz/Validation.scala#L37)
- [`WriterT#<++:`](https://github.com/scalaz/scalaz/blob/v7.1.0-M2/core/src/main/scala/scalaz/WriterT.scala#L36)
- [`Cord#+:`](https://github.com/scalaz/scalaz/blob/v7.1.0-M2/core/src/main/scala/scalaz/Cord.scala#L68)
- [`Cord#-:`](https://github.com/scalaz/scalaz/blob/v7.1.0-M2/core/src/main/scala/scalaz/Cord.scala#L74)
- [`EnumeratorT##::`](https://github.com/scalaz/scalaz/blob/v7.1.0-M2/iteratee/src/main/scala/scalaz/iteratee/EnumeratorT.scala#L17)

Should we provide alternative alias method and/or deprecate? rename?
Any good ideas?

**Comment by S11001001** [comment]
#1346

**Comment by S11001001** [comment]
Fixed by #1346.

**Comment by xuwei-k** [comment]
https://github.com/scala/scala/pull/5969
